### ISSUE TITLE
BG music issues

### ISSUE THREAD

**Issue by StraToN** [godotengine/escoria#35]
It is currently impossible to control the background music from Esc script.

Discussions from #34 
@fleskesvor 
>There's something strange going on with globals/bg_music.gd too, I think. It seems like set_state() is being called before this assignment happens: https://github.com/StraToN/escoria/blob/3.0/globals/bg_music.gd#L3
> which means the function returns on line 19, so it never complains about the errors on lines 28 and 30.

StraToN:
> I wonder if this is related to the fact that bg_music.tscn's script inherits item.gd.
>This causes this scene to be registered by the VM at launch although the scene is not ready yet. Maybe use call_deferred() ? Besides it causes the presence of an item in top-left corner of the game zone.
(telon.tscn has the same problem)

@fleskesvor 
> Yeah, that sounds plausible. The behavior is the exact same in the original Godot 2 version anyway, so this is not a regression. Does it cause any issues besides the fact that it's currently not possible to start background music from an .esc script? It's worth investigating, but maybe there should be an issue for it, since the problem is already present.

TL;DR
- bg_music.gd 's set_state() in called before the scene is properly instanced. Not sure if it is a bug in Godot.
- bg_music.gd and telon.gd inheriting item.gd may be a problem

**Comment by JuanNeedsToApologize** [comment]
Not a high priority at all. @reduz told me on reddit, jitter is completely normal for game engines.

**Comment by lavaduderDev** [comment]
This is seriously still an Issue? 
May I suggest making a dedicated function for changing the music. It'd be easier.

**Comment by mjtorn** [comment]
On Sat, Mar 10, 2018 at 11:00:52PM -0800, lavaduderDev wrote:
>This is seriously still an Issue? 
>May I suggest making a dedicated function for changing the music. It'd be easier.

I am nowhere near having music in anything, so I don't necessarily understand
this. I believe it's common to have an iMuse-like state machine for switching
between songs. This is not true in Escoria?

Tried looking around a bit for Wwise/Fmod for Godot, and apparently Fmod can
be integrated manually. Is this relevant for Escoria?

An explicit function definitely sounds like the easiest solution, so the game
can react to events.

**Comment by fleskesvor** [comment]
Godot has several nodes for playing audio, including [AudioStreamPlayer](http://docs.godotengine.org/en/3.0/classes/class_audiostreamplayer.html), which is suitable for background music, and [AudioStreamPlayer2D](http://docs.godotengine.org/en/3.0/classes/class_audiostreamplayer2d.html), which can be used for positional audio, like sound effects and speech.

There is bg_music.tscn, which can be used for setting background music, but currently there are some issues with setting music from .esc scripts.

The way it's supposed to work is that you call eg. `set bg_sound res://demo/audio/music/music.ogg` from an .esc script and then the `set_state` function of that node is called, since it's registered with the global_id `bg_music`.

This probably wasn't working in Godot 2.1 either, since the background music there was just added directly to the `bg_music.tscn` node and set on autoplay.

**Comment by mjtorn** [comment]
But basically you're saying that if things worked, triggering music changes on events would be doable?

What a shame it probably wasn't working. The world would be a better place if open-sourced projects actually worked :/ Maybe it's easy to fix.

**Comment by fleskesvor** [comment]
Yes, once that's fixed, it should just work. I will take another look at it now that the dust from the 3.0 migration has settled.

**Comment by Sorae** [comment]
Hi! The issue seems to be back in the `esc2-godot3.2` branch:
I'm not able to change the background music from an .esc scrip using set_state `bg_music res://xxx.ogg`, while it does work on the main branch.
